### PR TITLE
#163426519 implement signup feature

### PR DESCRIPTION
**Pull request by bobbylemm**
#### What does this PR do?
Create a controller for user sign up
#### Description of Task to be completed?
Have the following endpoints working

`post /api/v1/register`

#### How should this be manually tested?

After cloning the repo, cd into it and RUN `npm install` then `npm run dev`

- Using postman, go to the address`localhost:3000/api/v1`

- Provide a JSON object, that contains the following
`username`
`email`
`password`

#### What are the relevant pivotal tracker stories?
#163426519(https://www.pivotaltracker.com/story/show/163426519)

### CODE REVIEW

**Review by omenkish — DISMISSED** [review]

**Review by omenkish — DISMISSED** [review]

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by divinediscipline — DISMISSED** [review]
Looks good to me.

**Review by divinediscipline — COMMENTED** [review]
I noticed that the instructions on how to manually test this feature is not clear. Can you please review it to something along the lines of clone the repo, run "npm install" to install all dependencies, then run "npm run dev" to start the server. Using postman navigate to localhost:3000/api/v1/register and make a request ... Hope it makes sense.

File: server/controllers/userController.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import models from '../db/models';
+import assignToken from '../helpers/assignJwtToken';
+import errorMessage from '../helpers/errorHelpers';
+
+const { Users } = models;
+
+/**
+ *
+ *
+ * @class User
```
Comment (primuse) [inline server/controllers/userController.js]: In your JSDoc the name of the class is `user` but your actual class name is `UserController`. Please change this

File: server/controllers/userController.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ */
+class UserController {
+  /**
+   *
+   *
+   * @static
```
Comment (primuse) [inline server/controllers/userController.js]: Please add the name of the static method; `registerUser` in this case

**Review by omenkish — DISMISSED** [review]
lgtm

**Review by ozone4real — DISMISSED** [review]
LGTM

**Review by biolabalo — DISMISSED** [review]

**Review by primuse — DISMISSED** [review]

File: server/index.js [hunk 3]
```diff
@@ -1,81 +1,39 @@
...
-  res.status(err.status || 500);
-  res.json({
-    errors: {
-      message: err.message,
-      error: {}
-    }
-  });
-});
+// error handler
+app.use((err, req, res, next) => res.status(err.status).json({
```
[the remark this answers is not in the document]
Comment (p-funky) [reply server/index.js]: If next is never used then kindly remove it. 🙂

File: server/test/auth.test.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      .get('/*')
+      .then((res) => {
+        expect(res.body.error.message).to.equal('Not Found');
+        expect(res.status).to.equal(404);
+      });
+  });
+});
+
+describe('user registration', () => {
+  it('should fail on registration', () => request(app)
```
Comment (p-funky) [inline server/test/auth.test.js]: Why should this fail?

Comment (primuse) [reply server/test/auth.test.js]: @p-funky, that's because the `email` field wasn't filled and it is a required field

Comment (p-funky) [reply server/test/auth.test.js]: Yes, I thought as much. So it should be in the description of the test as well. Saying should `fail on registration` alone does not communicate fully.

Comment (bobbylemm) [reply server/test/auth.test.js]: ok, i will fix that now

**Review by omenkish — APPROVED** [review]

**Review by EfeAgare — APPROVED** [review]

**Review by divinediscipline — APPROVED** [review]
Looks good to me

**Review by biolabalo — APPROVED** [review]

**Review by primuse — APPROVED** [review]

**Review by p-funky — APPROVED** [review]

File: server/controllers/userController.js (hunk 1 above)
Comment (bobbylemm) [reply server/controllers/userController.js]: Thank you for the feedback, I have implemented this.

**Pull request merged**